### PR TITLE
HWKAPM-540 REST endpoint to return end to end Trace

### PR DESCRIPTION
**Pull request by objectiser**

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay Another PR for you to review when you get a chance. Main purpose of the PR is to provide a REST endpoint to obtain the complete end to end trace. This will be used to provide an end to end view of a trace instance in the UI. However also took the opportunity to do some refactoring and simplifying the TraceService API (and REST endpoints) to distinguish operations on trace and fragments.

File: client/trace-service-rest-client/src/main/java/org/hawkular/apm/trace/service/rest/client/TraceServiceRESTClient.java [hunk 1]
```diff
@@ -59,13 +59,70 @@ public TraceServiceRESTClient() {
...
+
+            is.close();
+
+            if (connection.getResponseCode() == 200) {
+                if (log.isLoggable(Level.FINEST)) {
+                    log.finest("Returned json=[" + new String(b) + "]");
+                }
+                try {
+                    return mapper.readValue(b, Trace.class);
+                } catch (Throwable t) {
```
Comment (pavolloffay) [inline client/trace-service-rest-client/src/main/java/org/hawkular/apm/trace/service/rest/client/TraceServiceRESTClient.java]: Isn't it better to catch more concrete ex?

File: client/trace-service-rest-client/src/main/java/org/hawkular/apm/trace/service/rest/client/TraceServiceRESTClient.java [hunk 2]
```diff
@@ -59,13 +59,70 @@ public TraceServiceRESTClient() {
...
+                    log.log(Level.SEVERE, "Failed to deserialize", t);
+                }
+            } else {
+                if (log.isLoggable(Level.FINEST)) {
+                    log.finest("Failed to get trace fragment: status=[" + connection.getResponseCode() + "]:"
+                            + connection.getResponseMessage());
+                }
+            }
+        } catch (Exception e) {
```
Comment (pavolloffay) [inline client/trace-service-rest-client/src/main/java/org/hawkular/apm/trace/service/rest/client/TraceServiceRESTClient.java]: Isn't it better to catch more concrete ex?

File: server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java [hunk 3]
```diff
@@ -96,18 +98,75 @@ public Trace get(String tenantId, String id) {
         }
 
         if (msgLog.isTraceEnabled()) {
```
Comment (pavolloffay) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java]: Are you testing the logging level for better performance?

Comment (objectiser) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java]: yes

File: client/trace-service-rest-client/src/main/java/org/hawkular/apm/trace/service/rest/client/TraceServiceRESTClient.java (hunk 1 above)
Comment (objectiser) [inline client/trace-service-rest-client/src/main/java/org/hawkular/apm/trace/service/rest/client/TraceServiceRESTClient.java]: will fix

File: client/trace-service-rest-client/src/main/java/org/hawkular/apm/trace/service/rest/client/TraceServiceRESTClient.java (hunk 2 above)
Comment (objectiser) [inline client/trace-service-rest-client/src/main/java/org/hawkular/apm/trace/service/rest/client/TraceServiceRESTClient.java]: will fix

**Comment by objectiser** [comment]
@pavolloffay Thanks!

**Pull request merged**